### PR TITLE
Fix __typename behavior and add missing subselections in result

### PR DESCRIPTION
**Pull request by gnawf**
This addresses the issues we've run into where `__typename` always resolves.

Also ensures there are no absent fields in the result. Fills in the absent fields with value of `null`.

Please make sure you consider the following:

- [x] Add tests that use __typename in queries
- [x] Does this change work with all nadel transformations (rename, type rename, hydration, etc)? Add tests for this.
- [x] Is it worth using hints for this change in order to be able to enable a percentage rollout?
- [x] Do we need to add integration tests for this change in the graphql gateway?
- [x] Do we need a pollinator check for this?

### CODE REVIEW

File: lib/src/main/java/graphql/nadel/NadelResultMerger.kt [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+            .errors(errors)
+            .build()
+    }
+
+    private fun fixData(
+        fields: List<NadelFieldAndService>,
+        engineSchema: GraphQLSchema,
+        data: MutableJsonMap,
+    ): MutableJsonMap? {
+        val requiredFieldMap = buildRequiredFieldMap(fields, engineSchema)
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/NadelResultMerger.kt:55]: So what this does is return a list of fields that should be in the result e.g. given a query

```graphql
query {
  projects {
    project(id: "1") {
      key
    }
    board(id: "1") {
      id
    }
  }
  issueById(id: "1") {
    key
  }
}
```

Then if `projects` is not null then we need both `project` and `board` in the result i.e. the following is a _valid_ result

```json
{
  "data": {
    "projects": {
      "project": {
        "key": "TEST"
      },
      "board": {
        "id": "1"
      }
    },
    "issueById": {
      "key": "TEST-1"
    }
  }
}
```

But, the following are not.

```json
{
  "data": {
    "projects": {
      "project": {
        "key": "TEST"
      }
    },
    "issueById": {
      "key": "TEST-1"
    }
  }
}
```

(above is missing `projects.board`)

```json
{
  "data": {
    "projects": {
      "board": {
        "id": "1"
      }
    },
    "issueById": {
      "key": "TEST-1"
    }
  }
}
```

(above is missing `projects.project`)

```json
{
  "data": {
    "projects": {
      "project": {
        "key": "TEST"
      },
      "board": {
        "id": "1"
      }
    }
  }
}
```

(above is missing `issueById`)

etc.

The `requiredFieldMap` for this query is

```json
{
  "projects": [
    "project",
    "board"
  ],
  "issueById": []
}
```

File: lib/src/main/java/graphql/nadel/NadelResultMerger.kt [hunk 2]
```diff
@@ -0,0 +1,156 @@
...
+
+            if (isNamespacedField(field, engineSchema)) {
+                requiredChildFields.addAll(field.children)
+            }
+        }
+
+        return requiredFields
+    }
+
+    private fun putAndMergeTopLevelData(data: MutableJsonMap, oneData: JsonMap) {
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/NadelResultMerger.kt:140]: This is pulled from GraphQLUtil.kt and is the same

File: lib/src/main/java/graphql/nadel/NextgenEngine.kt [hunk 3]
```diff
@@ -150,28 +149,29 @@ class NextgenEngine @JvmOverloads constructor(
             )
 
             val result: ExecutionResult = try {
-                mergeResults(
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/NextgenEngine.kt:153]: This change just moves the `mergeResults` to a function call below.

File: lib/src/main/java/graphql/nadel/NextgenEngine.kt [hunk 4]
```diff
@@ -296,12 +296,7 @@ class NextgenEngine @JvmOverloads constructor(
             )
         }
 
-        return serviceExecResult.copy(
-            data = serviceExecResult.data.let { data ->
-                data?.takeIf { transformedQuery.resultKey in data }
-                    ?: mutableMapOf(transformedQuery.resultKey to null)
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/NextgenEngine.kt:302]: This code was here to ensure that the result always has the top level field.

This has been moved to `NadelResultMerger` and now handles namespaced fields too.

File: lib/src/main/java/graphql/nadel/util/NamespacedUtil.kt [hunk 5]
```diff
@@ -18,4 +22,16 @@ object NamespacedUtil {
...
+        return overallField
+            .getFieldDefinitions(schema)
+            .any {
+                isNamespacedField(it)
+            }
+    }
+
+    fun isNamespacedField(definition: GraphQLFieldDefinition): Boolean {
+        return definition.hasAppliedDirective(namespacedDirectiveDefinition.name)
+    }
```
Comment (gnawf) [inline lib/src/main/java/graphql/nadel/util/NamespacedUtil.kt:36]: Moved from `NadelNamespacedFields`

**Review by temaEmelyan — APPROVED** [review]

**Pull request merged**